src/utils/nostr.ts

```diff
@@ -113,6 +121,90 @@ export const getNostrContacts = async (pubkey: string) => {
    return contactsPubkeys;
 };
 
+export const sendNip04DM = async (nprofile: string, content: string) => {
+   try {
+      const decoded = nip19.decode(nprofile);
+      if (decoded.type !== 'nprofile') {
+         throw new Error('give me an nprofile');
+      }
+      const { pubkey, relays } = decoded.data;
+      console.log('pubkey', pubkey);
+      console.log('relays', relays);
+      const { ndk, privkey } = await initializeNDK(relays);
+      const encryptedContent = await nip04.encrypt(privkey, pubkey, content);
+
+      const dmEvent = {
+         kind: NDKKind.EncryptedDirectMessage,
+         content: encryptedContent,
+         tags: [['p', pubkey]],
+      } as NostrEvent;
+
+      console.log('publishing event', dmEvent);
+      await publishNostrEvent(ndk, dmEvent, relays);
+   } catch (error) {
+      console.error('Error sending nostr dm', error);
+      throw new NostrError('');
+   }
+};
+
+const randomTimeUpTo2DaysInThePast = () => {
+   return Math.floor(Date.now() / 1000) - Math.floor(Math.random() * 172800);
+};
+
+export const sendNip17DM = async (nprofile: string, content: string) => {
+   try {
+      const decoded = nip19.decode(nprofile);
+      if (decoded.type !== 'nprofile') {
+         throw new Error('give me an nprofile');
+      }
+      const { pubkey: receiverPubkey, relays } = decoded.data;
+
+      console.log('sending DM to', receiverPubkey);
+      console.log('relays', relays);
+
+      const { ndk, privkey, pubkey: senderPubkey } = await initializeNDK(relays);
+
+      const msgEvent = new NDKEvent(undefined, {
+         kind: 14,
+         pubkey: senderPubkey,
+         created_at: Math.floor(Date.now() / 1000),
+         content,
+         tags: [['p', receiverPubkey]],
+      } as NostrEvent);
+      msgEvent.id = msgEvent.getEventHash();
+
+      const convoKey = nip44.getConversationKey(privkey, receiverPubkey);
+
+      const sealEvent = new NDKEvent(ndk, {
+         kind: 13,
+         tags: [],
+         pubkey: senderPubkey,
+         created_at: randomTimeUpTo2DaysInThePast(),
+         content: nip44.encrypt(JSON.stringify(msgEvent), convoKey),
+      } as NostrEvent);
+      await sealEvent.sign();
+
+      const { privkey: randPrivkey, pubkey: randPubkey } = generateKeyPair();
+      const giftConvoKey = nip44.getConversationKey(hexToBytes(randPrivkey), receiverPubkey);
+
+      const giftWrapEvent = new NDKEvent(ndk, {
+         kind: 1059,
+         pubkey: randPubkey,
+         created_at: randomTimeUpTo2DaysInThePast(),
```

Comment:
Privacy.

For DMs you are supposed to publish two events (one published to sender and another to receiver). I don't do this here because that is for the purpose of getting your conversation history. Having random timestamps makes it so you can't correlate those two events. That's the only reason I know of.

Reply:
its so that when you are looking for DMs sent to you now, you don't have to subscribe to all events in the past, just two days in the past